[benchmark][export] Add torch.export passrate for TB/TIMM benchmarks

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104382

issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test on TB and TIMM
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export --print-dataframe-summary
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export --print-dataframe-summary

why not HF
* huggingface use kwargs (dict) to torch.nn.module
* we will need to support kwargs in torch._export.export, which is in progress

local test result

timm 95% pass rate (58 ouf of 61 passed) P781702926
* 1 x [export specific]1 x ERROR:common:Mutating module attribute rel_indices during export 
* 1 x[not relevant to export] Unknown model (SelecSls42b)
* 1 x [not relevant to export] Failed to load model: HTTP Error 409: Public access is not permitted on this storage account

torchbench 54% pass rate (41 out of 75 passed) P781690552
* 7 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 3 x ERROR:common:call_method NNModuleVariable() / UserDefinedObjectVariable
* 3 x ERROR:common:Mutating module attribute {xx} during export.
* 2 x ERROR:common:inline in skipfiles
* 2 x ERROR:common:Consider annotating your code using constrain_as_*(). It appears that you're trying 
* 1 x ERROR:common:guard on data-dependent symbolic int/float
* 1 x ERROR:common:Tensor.tolist
* 1 x ERROR:common:Tensor.numpy. Turn on config.numpy_ndarray_as_tensor and install torch_np to support tensor.numpy().  [may be dev * env?]
* 1 x ERROR:common:missing: BUILD_SET
* 1 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:call_function BuiltinVariable(str) [GetAttrVariable(UserMethodVariable(<function 
* 1 x ERROR:common:Dynamic slicing on data-dependent value is not supported
  * 1 x ERROR:common:Failed running call_function <function interpolate at 0x7f60a8361ea0>(*(FakeTensor(..., device='cuda:0', size=(1, 3, * 427, 
* 1 x ERROR:common:Dynamo attempts to add additional input during export: value=0.6177528500556946, source=RandomValueSource(random_call_index=0)
* 1 x Found following user inputs located at [16, 17, 18, 19, 20, 21, 22] are mutated. This is currently banned in the aot_export workflow.
* 1 x RuntimeError: cumsum_cuda_kernel does not have a deterministic implementation
* 4 x pass_due_to_skip
* 1 x eager_2nd_run_OOM
* 1 x fail_accuracy


cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78

weifengpy committed:
[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks

issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result
* torchbench P779570524

[ghstack-poisoned]

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result
* torchbench 56% pass rate P779561366
* huggingface 0% pass rate P779605381
* timm TBD in progress

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result
* torchbench 56% pass rate P779561366
* huggingface 0% pass rate P779605381
* timm TBD in progress

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result
* torchbench 56% pass rate P779561366
* huggingface 0% pass rate P779605381
* timm 95% pass rate P779616493

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result

timm 75% pass rate P779992909
* 1 x ERROR:common:Mutating module attribute rel_indices during export 

huggingface 0% pass rate P779937906
* 23 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 15 x ERROR:common:Dynamo attempts to add additional input during export
* 3 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:It appears that you're trying to get a value out of symbolic 

torchbench 65% pass rate P779999869
* 7 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 3 x ERROR:common:call_method NNModuleVariable() / UserDefinedObjectVariable
* 2 x ERROR:common:Mutating module attribute {xx} during export.
* 2 x ERROR:common:inline in skipfiles
* 2 x ERROR:common:It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent...
* 1 x ERROR:common:guard on data-dependent symbolic int/float
* 1 x ERROR:common:Tensor.tolist
1 x ERROR:common:Tensor.numpy. Turn on config.numpy_ndarray_as_tensor and install torch_np to support tensor.numpy().  [may be dev * env?]
* 1 x ERROR:common:missing: BUILD_SET
* 1 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:call_function BuiltinVariable(str) [GetAttrVariable(UserMethodVariable(<function 
* 1 x ERROR:common:Dynamic slicing on data-dependent value is not supported
1 x ERROR:common:Failed running call_function <function interpolate at 0x7f60a8361ea0>(*(FakeTensor(..., device='cuda:0', size=(1, 3, * 427, 



cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

yanboliang commented:
Offline discussed with @willfengg:
* Correct the pass rate number, e.g, TIMM has 61 models and only one error, that pass rate should be 98%.
* Take a further look at HF pass rate, a bit wired. Check if applying ```torch.compile``` directly on top of models can make different, and figure out why HF benchmark has so many failures like ```Dynamo input and output is a strict subset of traced input/output```.

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result

timm 75% pass rate P779992909
* 1 x ERROR:common:Mutating module attribute rel_indices during export 

huggingface 0% pass rate P779937906
* 23 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 15 x ERROR:common:Dynamo attempts to add additional input during export
* 3 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:It appears that you're trying to get a value out of symbolic 

torchbench 65% pass rate P779999869
* 7 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 3 x ERROR:common:call_method NNModuleVariable() / UserDefinedObjectVariable
* 2 x ERROR:common:Mutating module attribute {xx} during export.
* 2 x ERROR:common:inline in skipfiles
* 2 x ERROR:common:It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent...
* 1 x ERROR:common:guard on data-dependent symbolic int/float
* 1 x ERROR:common:Tensor.tolist
1 x ERROR:common:Tensor.numpy. Turn on config.numpy_ndarray_as_tensor and install torch_np to support tensor.numpy().  [may be dev * env?]
* 1 x ERROR:common:missing: BUILD_SET
* 1 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:call_function BuiltinVariable(str) [GetAttrVariable(UserMethodVariable(<function 
* 1 x ERROR:common:Dynamic slicing on data-dependent value is not supported
1 x ERROR:common:Failed running call_function <function interpolate at 0x7f60a8361ea0>(*(FakeTensor(..., device='cuda:0', size=(1, 3, * 427, 



cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result

timm 75% pass rate P779992909
* 1 x ERROR:common:Mutating module attribute rel_indices during export 

huggingface 0% pass rate P779937906
* 23 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 15 x ERROR:common:Dynamo attempts to add additional input during export
* 3 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:It appears that you're trying to get a value out of symbolic 

torchbench 65% pass rate P779999869
* 7 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 3 x ERROR:common:call_method NNModuleVariable() / UserDefinedObjectVariable
* 2 x ERROR:common:Mutating module attribute {xx} during export.
* 2 x ERROR:common:inline in skipfiles
* 2 x ERROR:common:It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent...
* 1 x ERROR:common:guard on data-dependent symbolic int/float
* 1 x ERROR:common:Tensor.tolist
1 x ERROR:common:Tensor.numpy. Turn on config.numpy_ndarray_as_tensor and install torch_np to support tensor.numpy().  [may be dev * env?]
* 1 x ERROR:common:missing: BUILD_SET
* 1 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:call_function BuiltinVariable(str) [GetAttrVariable(UserMethodVariable(<function 
* 1 x ERROR:common:Dynamic slicing on data-dependent value is not supported
1 x ERROR:common:Failed running call_function <function interpolate at 0x7f60a8361ea0>(*(FakeTensor(..., device='cuda:0', size=(1, 3, * 427, 



cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result

timm 75% pass rate P779992909
* 1 x ERROR:common:Mutating module attribute rel_indices during export 

huggingface 0% pass rate P779937906
* 23 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 15 x ERROR:common:Dynamo attempts to add additional input during export
* 3 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:It appears that you're trying to get a value out of symbolic 

torchbench 65% pass rate P779999869
* 7 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 3 x ERROR:common:call_method NNModuleVariable() / UserDefinedObjectVariable
* 2 x ERROR:common:Mutating module attribute {xx} during export.
* 2 x ERROR:common:inline in skipfiles
* 2 x ERROR:common:It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent...
* 1 x ERROR:common:guard on data-dependent symbolic int/float
* 1 x ERROR:common:Tensor.tolist
1 x ERROR:common:Tensor.numpy. Turn on config.numpy_ndarray_as_tensor and install torch_np to support tensor.numpy().  [may be dev * env?]
* 1 x ERROR:common:missing: BUILD_SET
* 1 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:call_function BuiltinVariable(str) [GetAttrVariable(UserMethodVariable(<function 
* 1 x ERROR:common:Dynamic slicing on data-dependent value is not supported
1 x ERROR:common:Failed running call_function <function interpolate at 0x7f60a8361ea0>(*(FakeTensor(..., device='cuda:0', size=(1, 3, * 427, 



cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export

local test result

timm 95% pass rate (58 ouf of 61 passed) P780527230
* 1 x [export specific]1 x ERROR:common:Mutating module attribute rel_indices during export 
* 1 x[not relevant to export] Unknown model (SelecSls42b)
* 1 x [not relevant to export] Failed to load model: HTTP Error 409: Public access is not permitted on this storage account

huggingface 0% pass rate (46 in total) P780560251
* **[important]** 23 x ERROR:common:Dynamo input and output is a strict subset of traced input/output: unlike TIMM or TB, candidate_args is a subclass of transformers.modeling_outputs.ModelOutput instead of a tensor. **id(MaskedLMOutput) in dict_of_source_args will fail, but id(MaskedLMOutput.logits) can pass the assert**   code pointer: https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L884
  * 9 x arg=MaskedLMOutput
  * 9 x arg=QuestionAnsweringModelOutput
  * 4 x arg=CausalLMOutputWithCrossAttentions
  * 1 x arg=XLNetLMHeadModelOutput 
* 15 x ERROR:common:Dynamo attempts to add additional input during export
* 3 x ERROR:common:whole graph export entails exactly one guard export
* 2 x ERROR:common:call_function BuiltinVariable(setattr) [HFPretrainedConfigVariable(), ConstantVariable(str), ConstantVariable(str)] {}
* 1 x ERROR:common:It appears that you're trying to get a value out of symbolic 
* 2 x pass_due_to_skip

torchbench 57% pass rate (43 out of 75 passed) P780561084
* 7 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 3 x ERROR:common:call_method NNModuleVariable() / UserDefinedObjectVariable
* 3 x ERROR:common:Mutating module attribute {xx} during export.
* 2 x ERROR:common:inline in skipfiles
* 2 x ERROR:common:It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent...
* 1 x ERROR:common:guard on data-dependent symbolic int/float
* 1 x ERROR:common:Tensor.tolist
1 x ERROR:common:Tensor.numpy. Turn on config.numpy_ndarray_as_tensor and install torch_np to support tensor.numpy().  [may be dev * env?]
* 1 x ERROR:common:missing: BUILD_SET
* 1 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:call_function BuiltinVariable(str) [GetAttrVariable(UserMethodVariable(<function 
* 1 x ERROR:common:Dynamic slicing on data-dependent value is not supported
* 1 x ERROR:common:Failed running call_function <function interpolate at 0x7f60a8361ea0>(*(FakeTensor(..., device='cuda:0', size=(1, 3, * 427, 
* 1 x ERROR:common:Dynamo attempts to add additional input during export: value=0.6177528500556946, source=RandomValueSource(random_call_index=0)
* 1 x ERROR:common:Forward/backward hooks aren't yet supported on 'allowed' modules (e.g. <class 'torch.nn.modules.conv.Conv2d'>), which don't get traced through by dynamo. Graph-breaking to run hooks without compile.
* 4 x pass_due_to_skip
* 1 x eager_2nd_run_OOM


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

willfengg commented:
> Offline discussed with @willfengg:
> 
> * Correct the pass rate number, e.g, TIMM has 61 models and only one error, that pass rate should be 98%.
> * Take a further look at HF pass rate, a bit wired. Check if applying `torch.compile` directly on top of models can make different, and figure out why HF benchmark has so many failures like `Dynamo input and output is a strict subset of traced input/output`.

TIMM pass rate error is fixed and now it is 95% (58 ouf of 61 passed). I also updated TB and HF. The root cause for inaccurate pass error was because: I started TB/TIMM runs at the same time and they both write results to the same file export.csv. Now I run each one independently, they are good

for HF failure 'Dynamo input and output is a strict subset of traced input/output',  unlike TIMM or TB, candidate_args is a subclass of transformers.modeling_outputs.ModelOutput instead of a tensor. id(MaskedLMOutput) in dict_of_source_args will fail, but id(MaskedLMOutput.logits) can pass the assert code pointer: https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/eval_frame.py#L884

zhxchen17 commented on benchmarks/dynamo/common.py:
```diff
@@ -2970,6 +2998,17 @@ def run(runner, args, original_dir=None):
         output_filename = "inductor.csv"
+    elif args.export:
+        optimize_ctx = functools.partial(
+            torch._dynamo.export,
```
The correct API to use should be torch._export.export(). If you check the implementation it uses _dynamo.export internally as well and have extra logic to make sure we preserve all metadata correctly.

zhxchen17 reviewed (changes requested):
looks great and thanks for the work! some comment.

weifengpy committed:
Update on "[benchmark][export] Add torch.export passrate for TB/HF/TIMM benchmarks"


issues resolved: https://github.com/pytorch/pytorch/issues/104294

local test
* python benchmarks/dynamo/torchbench.py -d cuda --inference --accuracy --progress --export --print-dataframe-summary
* python benchmarks/dynamo/timm_models.py -d cuda --inference --accuracy --progress --export --print-dataframe-summary
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --export --print-dataframe-summary

local test result

timm 95% pass rate (58 ouf of 61 passed) P781702926
* 1 x [export specific]1 x ERROR:common:Mutating module attribute rel_indices during export 
* 1 x[not relevant to export] Unknown model (SelecSls42b)
* 1 x [not relevant to export] Failed to load model: HTTP Error 409: Public access is not permitted on this storage account

torchbench 54% pass rate (41 out of 75 passed) P781690552
* 7 x ERROR:common:Dynamo input and output is a strict subset of traced input/output
* 3 x ERROR:common:call_method NNModuleVariable() / UserDefinedObjectVariable
* 3 x ERROR:common:Mutating module attribute {xx} during export.
* 2 x ERROR:common:inline in skipfiles
* 2 x ERROR:common:Consider annotating your code using constrain_as_*(). It appears that you're trying 
* 1 x ERROR:common:guard on data-dependent symbolic int/float
* 1 x ERROR:common:Tensor.tolist
* 1 x ERROR:common:Tensor.numpy. Turn on config.numpy_ndarray_as_tensor and install torch_np to support tensor.numpy().  [may be dev * env?]
* 1 x ERROR:common:missing: BUILD_SET
* 1 x ERROR:common:whole graph export entails exactly one guard export
* 1 x ERROR:common:call_function BuiltinVariable(str) [GetAttrVariable(UserMethodVariable(<function 
* 1 x ERROR:common:Dynamic slicing on data-dependent value is not supported
  * 1 x ERROR:common:Failed running call_function <function interpolate at 0x7f60a8361ea0>(*(FakeTensor(..., device='cuda:0', size=(1, 3, * 427, 
* 1 x ERROR:common:Dynamo attempts to add additional input during export: value=0.6177528500556946, source=RandomValueSource(random_call_index=0)
* 1 x Found following user inputs located at [16, 17, 18, 19, 20, 21, 22] are mutated. This is currently banned in the aot_export workflow.
* 4 x pass_due_to_skip
* 1 x eager_2nd_run_OOM
* 1 x fail_accuracy


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

willfengg commented:
@pytorchbot merge

willfengg commented:
@pytorchbot label "topic: not user facing"

willfengg commented:
@pytorchbot merge
